compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt

```diff
@@ -0,0 +1,81 @@
+import org.jetbrains.kotlin.fir.expressions.FirGetClassCall
+import org.jetbrains.kotlin.fir.expressions.FirQualifiedAccessExpression
+import org.jetbrains.kotlin.fir.languageVersionSettings
+import org.jetbrains.kotlin.fir.references.FirResolvedNamedReference
+import org.jetbrains.kotlin.fir.resolve.containingClass
+
+object FirCallableReferenceChecker : FirQualifiedAccessChecker() {
+    override fun check(expression: FirQualifiedAccessExpression, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (expression !is FirCallableReferenceAccess) return
+        // [FirGetClassCallChecker] will check [FirGetClassCall].
+        if (expression is FirGetClassCall) return
+
+        checkReferenceIsToAllowedMember(expression, context, reporter)
+    }
+
+    // See FE 1.0 [DoubleColonExpressionResolver#checkReferenceIsToAllowedMember]
+    private fun checkReferenceIsToAllowedMember(
+        callableReferenceAccess: FirCallableReferenceAccess,
+        context: CheckerContext,
+        reporter: DiagnosticReporter
+    ) {
+        // UNRESOLVED_REFERENCE will be reported separately.
+        if (callableReferenceAccess.calleeReference !is FirResolvedNamedReference) return
+        val reference = callableReferenceAccess.calleeReference as FirResolvedNamedReference
+
+        val source = reference.source ?: return
+        if (source.kind is FirFakeSourceElementKind) return
+        // For example, rcv::member, reference source points to the entire callable reference.
+        // Note that both `rcv` and `member` would be of [REFERENCE_EXPRESSION] kind.
+        // We want to report on the last reference expression, i.e. the one after ::
+        val referenceSource = source.getLastChild(REFERENCE_EXPRESSION) ?: return
+
+        if (!context.session.languageVersionSettings.supportsFeature(LanguageFeature.CallableReferencesToClassMembersWithEmptyLHS)) {
+            if (callableReferenceAccess.explicitReceiver == null) {
+                reporter.reportOn(
+                    source,
+                    FirErrors.UNSUPPORTED_FEATURE,
+                    LanguageFeature.CallableReferencesToClassMembersWithEmptyLHS to context.session.languageVersionSettings,
+                    context
+                )
+            }
+        }
+        val referredDeclaration = reference.resolvedSymbol.fir
+        if (referredDeclaration is FirConstructor && referredDeclaration.containingClass?.classKind == ClassKind.ANNOTATION_CLASS) {
+            reporter.reportOn(referenceSource, FirErrors.CALLABLE_REFERENCE_TO_ANNOTATION_CONSTRUCTORS, context)
+        }
+        if (referredDeclaration is FirCallableMemberDeclaration<*> && referredDeclaration.isMemberExtension) {
+            reporter.reportOn(referenceSource, FirErrors.EXTENSION_IN_CLASS_REFERENCE_NOT_ALLOWED, referredDeclaration, context)
+        }
+        if (referredDeclaration is FirVariable<*> &&
+            referredDeclaration !is FirField &&
```

Comment:
You can have a reference to a field?

Can you try this:
```kt
val i: Int = 10
    get() {
        ::field  // Error in FE1.0: Unsupported [References to variables aren't supported yet]
        return field
    }
```

Reply:
It looks like there is no such test, thanks!

Btw, it's still `FirProperty` that such backing field references to, and yet it's not `local`. The check for `!is FirField` is there to discard Java field or delegate field, but seems a little bit vague. I rewrote here to clarify what we're reporting.

Reply:
Looks clearer, thanks!